optuna/artifacts/_upload.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+
+from dataclasses import asdict
+from dataclasses import dataclass
+import json
+import mimetypes
+import os
+import uuid
+
+from optuna.artifacts._protocol import ArtifactStore
+from optuna.storages import BaseStorage
+from optuna.trial._frozen import FrozenTrial
+from optuna.trial._trial import Trial
+
+
+ARTIFACTS_ATTR_PREFIX = "artifacts:"
+DEFAULT_MIME_TYPE = "application/octet-stream"
+
+
+@dataclass
+class ArtifactMeta:
+    artifact_id: str
+    filename: str
+    mimetype: str
+    encoding: str | None
+
+
+def upload_artifact(
+    trial: Trial | FrozenTrial | int,
```

Comment:
In Optuna, `trial_id` is actually not recommended to use directly. This is the reason why `trial_id` is private attributes in `Trial` and `FrozenTrial` classes and [Study.tell()](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.tell) accepts `trial_number`.

@HideakiImamura @gen740 Does it sound reasonable to change the interface like below?

```suggestion
    trial: Trial | FrozenTrial,
```

Reply:
Looks reasonable.

Reply:
I think too.